File: ios/NativeSigner/Modals/SufficientCryptoReady.swift
//
//  SufficientCryptoReady.swift
//  NativeSigner
//
//  Created by Alexander Slesarev on 21.12.2021.
//

import SwiftUI

struct SufficientCryptoReady: View {
    @GestureState private var dragOffset = CGSize.zero
    @State private var offset: CGFloat = 0
    @State private var oldOffset: CGFloat = 0
    var content: MSufficientCryptoReady
    var body: some View {
        ZStack {
            RoundedRectangle(cornerRadius: 8).foregroundColor(Asset.bg000.swiftUIColor)
            VStack {
                HeaderBar(line1: "Your Signature", line2: "Scan it into your application")
                Image(uiImage: UIImage(data: Data(content.sufficient)) ?? UIImage())
                    .resizable()
                    .aspectRatio(contentMode: .fit).padding(12)
                AddressCard(address: content.authorInfo)
                switch content.content {
                case let .addSpecs(network):
                    Text("Signature for network specs")
                    NetworkCard(title: network.networkTitle, logo: network.networkLogo)
                case .loadTypes(types: _, pic: _):
                    Text("Signature for types")
                case let .loadMetadata(name: name, version: version):
                    Text("Signature for metadata update")
                    Text(name + " version " + String(version))
                }
            }
        }
    }
}

// struct SufficientCryptoReady_Previews: PreviewProvider {
// static var previews: some View {
// SufficientCryptoReady()
// }
// }

File: android/src/main/java/io/parity/signer/modals/SufficientCryptoReady.kt
package io.parity.signer.modals

import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.*
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Surface
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.unit.dp
import io.parity.signer.components.HeaderBar
import io.parity.signer.components.Identicon
import io.parity.signer.components.KeyCard
import io.parity.signer.components.NetworkCard
import io.parity.signer.models.intoImageBitmap
import io.parity.signer.ui.theme.Bg000
import io.parity.signer.ui.theme.modal
import io.parity.signer.uniffi.MSufficientCryptoReady
import io.parity.signer.uniffi.MscContent

@Composable
fun SufficientCryptoReady(
	sufficientCrypto: MSufficientCryptoReady,
) {
	Surface(
		shape = MaterialTheme.shapes.modal,
		color = MaterialTheme.colors.Bg000
	) {
		Column(
			modifier = Modifier
				.fillMaxSize()
				.padding(20.dp)
		) {
			HeaderBar("Your signature", "Scan this into your application")
			Image(
				bitmap = sufficientCrypto.sufficient.intoImageBitmap(),
				contentDescription = "Signed update",
				contentScale = ContentScale.FillWidth,
				modifier = Modifier.fillMaxWidth()
			)
			KeyCard(
				identity = sufficientCrypto.authorInfo,
			)
			when (val c = sufficientCrypto.content) {
				is MscContent.AddSpecs -> Column {
					Text("Specs")
					NetworkCard(c.f)
				}
				is MscContent.LoadMetadata -> Text("Metadata for " + c.name + " with version " + c.version)
				is MscContent.LoadTypes -> Column {
					Text("types " + c.types)
					Identicon(identicon = c.pic)
				}
			}
		}
	}
}
